### DISCUSSION TITLE
Support MS-DOS. 2/5

### DISCUSSION THREAD

**Discussion by DosWorld** [samiam95124/Pascal-P6#63] [General]
I would ask you to add subj. This is not a joke. MS-DOS community still alive.

**Comment by samiam95124** [comment]
No I take it seriously. Rugxulo was doing work on Pascal-P5 for FreeDos using GPC, obviously with Pascal-P5 as an interpreter. That same method works for Pascal-P6, since it uses the same layout and is ISO 7185 compliant. The other pathway is to use the cmach interpreter, which bypasses GPC, and is faster to boot.

For compilers, it would need a 32 bit version of pgen, or is there a 64 bit extender I don't know about? I'll admit its not on the high priority list, since most major operating systems have moved on to 64 bits. My idea of priorities is for the AMD64 first, then ARM 64 and the Apple M series chips after that, simply by number of users.

I'll leave this ticket here. I think it needs a sponsor (you perhaps?).

**Comment by samiam95124** [comment]
So this is what I would do if I were going to port on DOS.

1. Pick a DOS, mostly likely freedos.
2. Compile cmach using your favorite compiler on DOS.
3. Fill out scripts to run programs with that, p6, compile and run. The .bat files are the starting point for that, but realize they may be out of date.
4. Place the result in a host directory like the gpc directory (freedos).
5. Create a configure script target to detect and enable that.
6. Run a full regression on the result.
7. Issue a pull request.

An alternative plan would be to translate pcom and pint to C via P5c, Trevor Blight's project. This is possible because P6 is built with ISO 7185.

Note that translation into C for pint is actually going to make it slower and harder to read. Why? cmach is essentially a stripped down version of pint for C that was hand programmed. P5c only really helps you if you translate everything you are going to run (ie, the target program) into C. When you translate pcom.pas into C, the running speed advantage is not that great, you get about an order of magnitude slower runs. Thus the advantage over running the whole thing in cmach is not as great as you would think.

**Comment by samiam95124** [comment]
Rated 2/5. I'll be frank, this ticket has two knocks against it, the first being that DOS is no longer mainstream. The second is that 32 bit mode on x86 is also no longer mainstream. This is different from (say) ARM, since that CPU has a large and very active embedded system base (I do ARM 32 bit embedded work, its a common bluetooth/wifi processor).

None of this means it is a bad target, it is simply that I have to chose things to work on based on the number of potential users first. The other issue is that, even though I used to do a lot of work on DOS, its been decades since I did that kind of work, so I would not be the first choice of people to work on this.

This means that it would be the perfect project for someone who has a strong interest and skills in DOS, likely FreeDOS, to work on. The fact that Pascal-P6's pgen module 64 bit only is not necessarily a blocker, since there is a 32 bit DOS version of GPC and cmach can be configured as 32 bit. In addition, if the issue had a sponsor to work on it that might be the thing that would convince me to produce a 32 bit pgen module.

Regards

**Comment by samiam95124** [comment]
Since there is no way I am going to do this ticket, for reasons explained above, I am kicking this issue to discussions. This could change if someone wants to pick it up, but I feel issues should be left for items that have a plan to complete them.
